File: examples/go/log_filter.go
// log_filter.go — Go equivalent of examples/log_filter.cnd
//
// What the signature CANNOT tell you that Candor signatures CAN:
//   - parseLevel, keepLine, filterLines: pure? the signature is silent
//   - runFilter: has I/O? (string, string, string) -> (int, error) gives no hint
//   - the int return from runFilter can be silently discarded: runFilter(a, b, c)
//
// In Candor:
//   fn parseLevel(line: str) -> option<str> pure   ← pure: enforced, not a comment
//   fn runFilter(in, out, level: str) -> result<i64, str> effects(io)

package main

import (
	"fmt"
	"os"
	"strings"
)

func parseLevel(line string) (string, bool) {
	if len(line) < 3 || line[0] != '[' {
		return "", false
	}
	close := strings.Index(line[1:], "]")
	if close < 0 {
		return "", false
	}
	return line[1 : close+1], true
}

func keepLine(line, level string) bool {
	lvl, ok := parseLevel(line)
	if !ok {
		return false
	}
	return lvl == level
}

func filterLines(lines []string, level string) []string {
	var out []string
	for _, line := range lines {
		if keepLine(line, level) {
			out = append(out, line)
		}
	}
	return out
}

func runFilter(pathIn, pathOut, level string) (int, error) {
	data, err := os.ReadFile(pathIn)
	if err != nil {
		return 0, fmt.Errorf("cannot read: %s", pathIn)
	}
	lines := strings.Split(string(data), "\n")
	kept := filterLines(lines, level)
	out := strings.Join(kept, "\n") + "\n"
	if err := os.WriteFile(pathOut, []byte(out), 0644); err != nil {
		return 0, fmt.Errorf("cannot write: %s", err)
	}
	return len(kept), nil
}

func main() {
	log := "[ERROR] disk full\n[INFO] server started\n[WARN] high memory\n[ERROR] connection refused\n[INFO] request received\n"
	if err := os.WriteFile("examples/sample.log", []byte(log), 0644); err != nil {
		fmt.Println("setup error:", err)
		return
	}
	n, err := runFilter("examples/sample.log", "examples/errors.log", "ERROR")
	if err != nil {
		fmt.Println("filter failed:", err)
		return
	}
	fmt.Printf("%d ERROR lines written to errors.log\n", n)
}

File: examples/rust/log_filter.rs
// log_filter.rs — Rust equivalent of examples/log_filter.cnd
//
// What the signature CANNOT tell you that Candor signatures CAN:
//   - parse_level, keep_line, filter_lines: pure? no Rust annotation for runtime purity
//   - run_filter: the signature Result<i64, String> says "can fail" but not "does I/O"
//     An AI reviewing this signature must read the body to know it touches the filesystem
//   - Result is enforced; purity and effects are not declared anywhere
//
// In Candor:
//   fn parse_level(line: str) -> option<str> pure   ← pure: compiler-verified
//   fn run_filter(in, out, level: str) -> result<i64, str> effects(io)

use std::fs;

fn parse_level(line: &str) -> Option<&str> {
    if line.len() < 3 || !line.starts_with('[') {
        return None;
    }
    let close = line[1..].find(']')?;
    Some(&line[1..close + 1])
}

fn keep_line(line: &str, level: &str) -> bool {
    parse_level(line).map_or(false, |lvl| lvl == level)
}

fn filter_lines<'a>(lines: &[&'a str], level: &str) -> Vec<&'a str> {
    lines.iter().filter(|&&l| keep_line(l, level)).copied().collect()
}

fn run_filter(path_in: &str, path_out: &str, level: &str) -> Result<i64, String> {
    let raw = fs::read_to_string(path_in)
        .map_err(|_| format!("cannot read: {}", path_in))?;
    let lines: Vec<&str> = raw.lines().collect();
    let kept = filter_lines(&lines, level);
    let out = kept.join("\n") + "\n";
    fs::write(path_out, out)
        .map_err(|e| format!("cannot write: {}", e))?;
    Ok(kept.len() as i64)
}

fn main() {
    let log = "[ERROR] disk full\n[INFO] server started\n[WARN] high memory\n[ERROR] connection refused\n[INFO] request received\n";
    fs::write("examples/sample.log", log).expect("setup error");
    match run_filter("examples/sample.log", "examples/errors.log", "ERROR") {
        Ok(n)  => println!("{} ERROR lines written to errors.log", n),
        Err(e) => println!("filter failed: {}", e),
    }
}
